### PR TITLE
ctc grad scale same with espnet

### PR DESCRIPTION
**Pull request by zh794390558**
### PR types
New features

### PR changes
OPs

### Describe

1. CTC 算 loss有两种norm方式，一种是 batch_size 的 reduce，一种是 element_wise 的reduce。
这个PR是修正这两种 norm 方式下 ctc grad 的计算。

2. 目前 paddle 的 norm 方式和 espnet 是有差异的，而且现在实现的方式并不常用。
所以新增两种 norm 方式，具体实现同  https://github.com/espnet/warp-ctc/blob/pytorch_bindings/pytorch_binding/warpctc_pytorch/__init__.py#L39

3. 具体使用时需要用户自己根据使用的norm方式选在对应的选项对 grad 做处理。

4. 修改文件过多是因为 ctc 实现了 LodTensor。
5. doc https://github.com/PaddlePaddle/docs/pull/3709
6. grad支持GPU操作

espnet:
image

paddle:
image

paddle-gpu:
image

paddle-cpu vs paddle-gpu
测试case和espnet不同
ctc_grad_gpu_pr

### CODE REVIEW

**Review by Shixiaowei02 — DISMISSED** [review]

File: paddle/fluid/operators/warpctc_op.h [hunk 1]
```diff
@@ -366,22 +367,49 @@ class WarpCTCGradKernel : public framework::OpKernel<T> {
...
+          "[warpctc grad] norm_by_times, size_average and length_average "
+          "should one be true."));
+    }
 
     if (ctx.HasInput("LogitsLength")) {
-      size_t max_seq_length = warpctc_grad->dims()[0];
-      size_t num_sequences = warpctc_grad->dims()[1];
-      size_t seq_width = warpctc_grad->dims()[2];
+      size_t Tmax = warpctc_grad->dims()[0];
+      size_t B = warpctc_grad->dims()[1];
```
Comment (ZeyuChen) [inline paddle/fluid/operators/warpctc_op.h]: 这一参数量的变动反而降低的可读性，建议还原。

File: paddle/fluid/operators/math/sequence_padding.cc [hunk 2]
```diff
@@ -54,7 +55,21 @@ void CopyValidData(framework::Tensor* dst_tensor,
     int pad_data_offset = layout == kBatchLengthWidth
                               ? seq_idx * pad_seq_len * step_width
                               : seq_idx * step_width;
-    float scale = 1.0f / static_cast<float>(valid_seq_len);
+
+    float scale = 1.0f;
+    if (length_average) {
```
Comment (ZeyuChen) [inline paddle/fluid/operators/math/sequence_padding.cc]: 感觉改为norm_by_xxx比较合适

File: paddle/fluid/operators/warpctc_op.h [hunk 3]
```diff
@@ -366,22 +367,49 @@ class WarpCTCGradKernel : public framework::OpKernel<T> {
...
                             &logits_length_cpu);
 
+      // total logits length
+      int total_length = 0;
+      auto* length_ptr = logits_length_cpu.data<int64_t>();
+      for (size_t i = 0; i < B; i++) {
+        total_length += length_ptr[i];
+      }
+      VLOG(3) << "[warpctc grad] total logits length: " << total_length;
+      if (length_average) {
```
Comment (ZeyuChen) [inline paddle/fluid/operators/warpctc_op.h]: 统一到原有实现中nomr_by_xxx的命名体系中，避免代码风格割裂。

File: paddle/fluid/operators/warpctc_op.h [hunk 4]
```diff
@@ -397,25 +425,34 @@ class WarpCTCGradKernel : public framework::OpKernel<T> {
...
 
       const T* loss_grad_data = loss_grad_cpu.data<T>();
-      for (size_t i = 0; i < max_seq_length; ++i) {
-        for (size_t j = 0; j < num_sequences; ++j) {
+      for (size_t i = 0; i < Tmax; ++i) {
+        for (size_t j = 0; j < B; ++j) {
           T scale = 1.0;
-          if (norm_by_times) {
-            scale = 1.0 / static_cast<T>(logits_length_cpu.data<int64_t>()[j]);
+          if (length_average) {
```
Comment (ZeyuChen) [inline paddle/fluid/operators/warpctc_op.h]: 统一到原有实现中nomr_by_xxx的命名体系中，避免代码风格割裂。

File: paddle/fluid/operators/warpctc_op.h [hunk 5]
```diff
@@ -397,25 +425,34 @@ class WarpCTCGradKernel : public framework::OpKernel<T> {
...
+            // Compute the avg. log-probability per batch sample and frame.
+            // https://github.com/espnet/warp-ctc/blob/pytorch_bindings/pytorch_binding/warpctc_pytorch/__init__.py#L42
+            scale = 1.0 / static_cast<T>(total_length);
+          } else if (size_average) {
```
Comment (ZeyuChen) [inline paddle/fluid/operators/warpctc_op.h]: 统一到原有实现中nomr_by_xxx的命名体系中，避免代码风格割裂。

Comment (ZeyuChen) [reply paddle/fluid/operators/warpctc_op.h]: 完善注释说明这一CTC计算行为与ESPNet保持一致。

File: python/paddle/fluid/layers/loss.py [hunk 6]
```diff
@@ -478,7 +478,9 @@ def warpctc(input,
             blank=0,
             norm_by_times=False,
             input_length=None,
-            label_length=None):
+            label_length=None,
+            size_average=False,
```
Comment (ZeyuChen) [inline python/paddle/fluid/layers/loss.py]: 命名风格统一为norm_by_xxx

**Review by ZeyuChen — COMMENTED** [review]
与ESPNet行为进一步验证，命名风格需统一。

File: python/paddle/fluid/tests/unittests/op_test.py [hunk 7]
```diff
@@ -241,6 +241,30 @@ def wrapper(cls):
     return wrapper
 
 
+def skip_check_grad_assert_ci(reason=None):
```
Comment (XiaoguangHu01) [inline python/paddle/fluid/tests/unittests/op_test.py]: 这里新增一个装饰器的原因是什么？看起来是共性的需求，是否可以复用skip_check_grad_ci
https://github.com/PaddlePaddle/Paddle/wiki/Gradient-Check-Is-Required-for-Op-Test

File: paddle/fluid/operators/math/sequence_padding.h [hunk 8]
```diff
@@ -117,6 +119,8 @@ class UnpaddingLoDTensorFunctor {
                   const framework::LoDTensor& pad_tensor,
                   framework::LoDTensor* seq_tensor, int pad_seq_len = -1,
                   int lod_level = 0, bool norm_by_times = false,
+                  bool norm_by_batchsize = false,
+                  bool norm_by_total_logits_len = false,
```
Comment (XiaoguangHu01) [inline paddle/fluid/operators/math/sequence_padding.h:123]: 从兼容性考虑，建议将这两个参数放到最后，避免使用了layout的参数调用失败。

File: paddle/fluid/operators/math/sequence_padding.cu [hunk 9]
```diff
@@ -118,6 +130,8 @@ class UnpaddingLoDTensorFunctor<platform::CUDADeviceContext, T> {
                   const framework::LoDTensor& pad_tensor,
                   framework::LoDTensor* seq_tensor, int pad_seq_len = -1,
                   int lod_level = 0, bool norm_by_times = false,
+                  bool norm_by_batchsize = false,
+                  bool norm_by_total_logits_len = false,
```
Comment (XiaoguangHu01) [inline paddle/fluid/operators/math/sequence_padding.cu:134]: 从兼容性考虑，建议将这两个参数放到最后，避免使用了layout的参数调用失败。

File: paddle/fluid/operators/math/sequence_padding.cu [hunk 10]
```diff
@@ -107,7 +119,7 @@ class PaddingLoDTensorFunctor<platform::CUDADeviceContext, T> {
     SequencePaddingKernel<T, kSeqToPad><<<grid, threads, 0, context.stream()>>>(
         pad_data, seq_data, pad_value_data, pad_value.numel() == 1,
         seq_offsets.CUDAData(context.GetPlace()), seq_num, pad_seq_len,
-        step_width, norm_by_times, layout);
+        step_width, norm_by_times, false, false, 0, layout);
```
Comment (XiaoguangHu01) [inline paddle/fluid/operators/math/sequence_padding.cu:122]: 参数0是指什么？

File: paddle/fluid/operators/math/sequence_padding.cc [hunk 11]
```diff
@@ -54,7 +55,21 @@ void CopyValidData(framework::Tensor* dst_tensor,
     int pad_data_offset = layout == kBatchLengthWidth
                               ? seq_idx * pad_seq_len * step_width
                               : seq_idx * step_width;
-    float scale = 1.0f / static_cast<float>(valid_seq_len);
+
+    float scale = 1.0f;
```
Comment (XiaoguangHu01) [inline paddle/fluid/operators/math/sequence_padding.cc:59]: 为啥需要在CopyValidData做这个操作呢？
我看在WarpCTCGradKernel 也做了一遍类似的计算，这两者的关系是什么？

Comment (zh794390558) [reply paddle/fluid/operators/math/sequence_padding.cc:59]: 这里的LodTensor实现中加入次操作，WarpCTCGradKernel里加的是Tensor的操作。

File: python/paddle/fluid/tests/unittests/op_test.py (hunk 7 above)
Comment (zh794390558) [reply python/paddle/fluid/tests/unittests/op_test.py]: 这里加这个是为了跑grad 的coverage，但是不去对grad的数据做assert的推断，因为用第三方的库无法准确的计算grad。

File: paddle/fluid/operators/math/sequence_padding.cu (hunk 10 above)
Comment (zh794390558) [reply paddle/fluid/operators/math/sequence_padding.cu:122]: 零是 total_logits_len， norm_by_total_logits_len的时候需要用次参数。

File: paddle/fluid/operators/math/sequence_padding.h (hunk 8 above)
Comment (zh794390558) [reply paddle/fluid/operators/math/sequence_padding.h:123]: 所有相关的操作都替换了，目前不会有兼容性问题。

File: paddle/fluid/operators/math/sequence_padding.cu (hunk 9 above)
Comment (zh794390558) [reply paddle/fluid/operators/math/sequence_padding.cu:134]: 所有相关的操作都替换了，目前不会有兼容性问题。

**Review by zhangting2020 — APPROVED** [review]
LGTM for skip_check_grad_ci

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Review by ZeyuChen — APPROVED** [review]
LGTM
